pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+
+	return allErrs
+}
+
+// ValidateFlowSchemaStatus validates status for the flow-schema.
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
+	return allErrs
+}
+
+// ValidatePriorityLevelConfiguration validates priority-level-configuration.
+func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationSpec(&pl.Spec, pl.Name, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationStatus(&pl.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidatePriorityLevelConfigurationSpec validates priority-level-configuration's spec.
+func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfigurationSpec, name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch spec.Type {
+	case flowcontrol.PriorityLevelQueuingTypeExempt:
+		if name != flowcontrol.PriorityLevelConfigurationNameExempt {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("type"), fmt.Sprintf(`must be named "exempt" to be exempt queuing type`)))
+		}
+	case flowcontrol.PriorityLevelQueuingTypeQueueing:
+		allErrs = append(allErrs, ValidatePriorityLevelQueuingConfiguration(spec.Queuing, fldPath.Child("queuing"))...)
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("type"), spec.Type, supportedQueuingType.List()))
+	}
+	return allErrs
+}
+
+// ValidatePriorityLevelQueuingConfiguration validates queuing-configuration for a priority-level
+func ValidatePriorityLevelQueuingConfiguration(queuing *flowcontrol.QueuingConfiguration, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if queuing.AssuredConcurrencyShares < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("assuredConcurrencyShares"), queuing.AssuredConcurrencyShares, "must be positive"))
+	}
+	if queuing.QueueLengthLimit < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queueLengthLimit"), queuing.QueueLengthLimit, "must be positive"))
+	}
+
+	// validate input arguments for shuffle-sharding
+	if queuing.Queues < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queues"), queuing.Queues, "must be positive"))
+	}
+	if queuing.HandSize < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("handSize"), queuing.HandSize, "must be positive"))
+	}
+	if queuing.HandSize > queuing.Queues {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("handSize"), queuing.HandSize,
+			fmt.Sprintf("should not be greater than queues (%d)", queuing.Queues)))
+	}
+	if entropy := shufflesharding.RequiredEntropyBits(int(queuing.Queues), int(queuing.HandSize)); entropy > shufflesharding.MaxHashBits {
```

Comment:
A call to `shufflesharding.RequiredEntropyBits` needs to be preceded by a guarantee that the deck size is `<= 1<<26`.

Reply:
Indeed, we should feel free to set even more restrictive limits in this validation code. Need it ever be higher than, say, 10000? 100000?

Reply:
I think the validation should match the documentation.

I do not want to bikeshed limits here.  We need a limit of `1<<26` to avoid overflow in `shufflesharding.RequiredEntropyBits`.  Bringing in other issues would just open up more things to argue about, and we will probably settle on putting a discussion in the user-facing documentation.  Let's not worry about establishing lower limits here now.

Reply:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas it says we can neither relax or restrict validations for released APIs, but it's fine for alpha API, is it?

Reply:
Validations here should be a subset of validations in shuffle sharding, so I think 10^7 or even more restrictive limits is OK. IMO, 10^5 is enough and also have good readability.

Reply:
@yue9944882: yes, it looks to me like we can change validation on fields in an alpha API.